File: migrate_csdn.js
#!/usr/bin/env node
/**
 * CSDN to Hexo Blog Migration Script
 * Extracts article info from CSDN HTML listing and creates Hexo markdown posts.
 * 
 * Usage:
 *   node migrate_csdn.js                    # Extract metadata only (fast)
 *   node migrate_csdn.js --fetch-content    # Also fetch full content from CSDN
 */

const fs = require('fs');
const path = require('path');
const https = require('https');
const http = require('http');

// Configuration
const CONFIG = {
  inputFile: 'data',
  outputDir: 'source/_posts',
  articlesJson: 'articles.json',
  delayBetweenRequests: 2000, // ms
};

/**
 * Parse HTML to extract article listings
 */
function parseArticleList(html) {
  const articles = [];
  
  // Match each article block
  const articleRegex = /<article[^>]*class="blog-list-box"[^>]*>([\s\S]*?)<\/article>/gi;
  let match;
  
  while ((match = articleRegex.exec(html)) !== null) {
    const block = match[1];
    const article = {};
    
    // Extract URL and ID
    const urlMatch = block.match(/href="(https:\/\/blog\.csdn\.net\/[^"]*\/article\/details\/(\d+))"/);
    if (urlMatch) {
      article.url = urlMatch[1];
      article.id = urlMatch[2];
    }
    
    // Extract title
    const titleMatch = block.match(/<h4[^>]*>([\s\S]*?)<\/h4>/i);
    if (titleMatch) {
      article.title = cleanText(titleMatch[1]);
    }
    
    // Extract description
    const descMatch = block.match(/class="blog-list-content"[^>]*>([\s\S]*?)<\/div>/i);
    if (descMatch) {
      article.description = cleanText(descMatch[1]);
    }
    
    // Extract date (format: 发布博客&nbsp;YYYY.MM.DD)
    const dateMatch = block.match(/发布博客[^0-9]*(\d{4})\.(\d{2})\.(\d{2})/);
    if (dateMatch) {
      article.date = `${dateMatch[1]}-${dateMatch[2]}-${dateMatch[3]}`;
    }
    
    // Extract views
    const viewsMatch = block.match(/class="view-num"[^>]*>(\d+)/);
    if (viewsMatch) {
      article.views = parseInt(viewsMatch[1], 10);
    }
    
    // Extract likes
    const likesMatch = block.match(/class="give-like-num"[^>]*>(\d+)/);
    if (likesMatch) {
      article.likes = parseInt(likesMatch[1], 10);
    }
    
    // Extract comments and favorites from comment-box
    const commentMatches = block.match(/class="comment-num"[^>]*>(\d+)/g);
    if (commentMatches && commentMatches.length >= 1) {
      const commentsNum = commentMatches[0].match(/(\d+)/);
      if (commentsNum) article.comments = parseInt(commentsNum[1], 10);
    }
    if (commentMatches && commentMatches.length >= 2) {
      const favoritesNum = commentMatches[1].match(/(\d+)/);
      if (favoritesNum) article.favorites = parseInt(favoritesNum[1], 10);
    }
    
    // Extract cover image
    const coverMatch = block.match(/class="course-img"[^>]*src="([^"]+)"/);
    if (coverMatch) {
      article.cover = coverMatch[1];
    }
    
    if (article.url && article.title) {
      articles.push(article);
    }
  }
  
  return articles;
}

/**
 * Clean HTML text
 */
function cleanText(html) {
  return html
    .replace(/<[^>]+>/g, '')
    .replace(/&nbsp;/g, ' ')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&amp;/g, '&')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/\s+/g, ' ')
    .trim();
}

/**
 * Sanitize filename
 */
function sanitizeFilename(title) {
  return title
    .replace(/[<>:"/\\|?*]/g, '')
    .replace(/\s+/g, '-')
    .replace(/-+/g, '-')
    .substring(0, 100)
    .replace(/^-|-$/g, '');
}


/**
 * Fetch article content from CSDN
 */
function fetchArticleContent(url) {
  return new Promise((resolve, reject) => {
    const options = {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        'Accept': 'text/html,application/xhtml+xml',
        'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8',
        'Referer': 'https://blog.csdn.net/',
      },
    };
    
    const protocol = url.startsWith('https') ? https : http;
    
    protocol.get(url, options, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        // Follow redirect
        fetchArticleContent(res.headers.location).then(resolve).catch(reject);
        return;
      }
      
      let data = '';
      res.setEncoding('utf8');
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const result = parseArticleContent(data);
          resolve(result);
        } catch (e) {
          reject(e);
        }
      });
    }).on('error', reject);
  });
}

/**
 * Parse article page content
 */
function parseArticleContent(html) {
  let content = '';
  let tags = [];
  let category = 'General';
  
  // Extract main content
  const contentMatch = html.match(/id="content_views"[^>]*>([\s\S]*?)<\/div>\s*<\/article>/i) ||
                       html.match(/class="article_content"[^>]*>([\s\S]*?)<\/div>\s*<\/article>/i);
  
  if (contentMatch) {
    content = convertHtmlToMarkdown(contentMatch[1]);
  }
  
  // Extract tags
  const tagMatches = html.matchAll(/class="tag-link"[^>]*>([^<]+)<\/a>/gi);
  for (const m of tagMatches) {
    const tag = cleanText(m[1]);
    if (tag && !tags.includes(tag)) {
      tags.push(tag);
    }
  }
  
  // Extract category from column name
  const categoryMatch = html.match(/class="column_name"[^>]*>([^<]+)<\/a>/i);
  if (categoryMatch) {
    const cat = cleanText(categoryMatch[1]);
    if (cat && cat !== '原创' && cat !== '转载') {
      category = cat;
    }
  }
  
  return { content, tags, category };
}

/**
 * Convert HTML to Markdown
 */
function convertHtmlToMarkdown(html) {
  let md = html;
  
  // Remove script and style
  md = md.replace(/<script[\s\S]*?<\/script>/gi, '');
  md = md.replace(/<style[\s\S]*?<\/style>/gi, '');
  
  // Code blocks
  md = md.replace(/<pre[^>]*><code[^>]*class="[^"]*(?:language-|lang-)(\w+)[^"]*"[^>]*>([\s\S]*?)<\/code><\/pre>/gi, 
    (_, lang, code) => `\n\`\`\`${lang}\n${cleanText(code)}\n\`\`\`\n`);
  md = md.replace(/<pre[^>]*><code[^>]*>([\s\S]*?)<\/code><\/pre>/gi, 
    (_, code) => `\n\`\`\`\n${cleanText(code)}\n\`\`\`\n`);
  md = md.replace(/<pre[^>]*>([\s\S]*?)<\/pre>/gi, 
    (_, code) => `\n\`\`\`\n${cleanText(code)}\n\`\`\`\n`);
  
  // Inline code
  md = md.replace(/<code[^>]*>([^<]*)<\/code>/gi, '`$1`');
  
  // Headers
  for (let i = 6; i >= 1; i--) {
    const regex = new RegExp(`<h${i}[^>]*>([\\s\\S]*?)<\\/h${i}>`, 'gi');
    md = md.replace(regex, (_, text) => `\n${'#'.repeat(i)} ${cleanText(text)}\n`);
  }
  
  // Bold and italic
  md = md.replace(/<(?:b|strong)[^>]*>([\s\S]*?)<\/(?:b|strong)>/gi, '**$1**');
  md = md.replace(/<(?:i|em)[^>]*>([\s\S]*?)<\/(?:i|em)>/gi, '*$1*');
  
  // Links
  md = md.replace(/<a[^>]*href="([^"]*)"[^>]*>([^<]*)<\/a>/gi, '[$2]($1)');
  
  // Images
  md = md.replace(/<img[^>]*src="([^"]*)"[^>]*alt="([^"]*)"[^>]*\/?>/gi, '\n![$2]($1)\n');
  md = md.replace(/<img[^>]*src="([^"]*)"[^>]*\/?>/gi, '\n![image]($1)\n');
  
  // Lists
  md = md.replace(/<li[^>]*>([\s\S]*?)<\/li>/gi, (_, text) => `- ${cleanText(text)}\n`);
  md = md.replace(/<\/?[uo]l[^>]*>/gi, '\n');
  
  // Blockquotes
  md = md.replace(/<blockquote[^>]*>([\s\S]*?)<\/blockquote>/gi, 
    (_, text) => `\n> ${cleanText(text)}\n`);
  
  // Paragraphs and line breaks
  md = md.replace(/<br\s*\/?>/gi, '\n');
  md = md.replace(/<p[^>]*>([\s\S]*?)<\/p>/gi, '\n$1\n');
  md = md.replace(/<div[^>]*>([\s\S]*?)<\/div>/gi, '\n$1\n');
  
  // Remove remaining HTML tags
  md = md.replace(/<[^>]+>/g, '');
  
  // Clean up entities
  md = md.replace(/&nbsp;/g, ' ');
  md = md.replace(/&lt;/g, '<');
  md = md.replace(/&gt;/g, '>');
  md = md.replace(/&amp;/g, '&');
  md = md.replace(/&quot;/g, '"');
  
  // Clean up whitespace
  md = md.replace(/\n{3,}/g, '\n\n');
  
  return md.trim();
}


/**
 * Create Hexo post content
 */
function createHexoPost(article, contentData = {}) {
  const title = (article.title || 'Untitled').replace(/"/g, '\\"');
  const date = article.date || new Date().toISOString().split('T')[0];
  const category = contentData.category || 'General';
  const tags = contentData.tags || [];
  const content = contentData.content || article.description || '';
  
  let frontMatter = `---
title: "${title}"
date: ${date}
updated: ${date}
categories:
  - ${category}
`;

  if (tags.length > 0) {
    frontMatter += 'tags:\n';
    tags.forEach(tag => {
      frontMatter += `  - ${tag}\n`;
    });
  }

  frontMatter += `csdn_views: ${article.views || 0}
csdn_likes: ${article.likes || 0}
csdn_comments: ${article.comments || 0}
csdn_favorites: ${article.favorites || 0}
csdn_url: ${article.url || ''}
`;

  if (article.cover) {
    frontMatter += `cover: ${article.cover}\n`;
  }

  if (article.description) {
    const desc = article.description.substring(0, 200).replace(/"/g, '\\"');
    frontMatter += `description: "${desc}"\n`;
  }

  frontMatter += '---\n\n';

  // Add original link notice
  const originalNotice = `> 本文迁移自CSDN博客，原文链接：[${article.title}](${article.url})
> 
> 原文数据：👀 ${article.views || 0} 阅读 | 👍 ${article.likes || 0} 点赞 | 💬 ${article.comments || 0} 评论 | ⭐ ${article.favorites || 0} 收藏

`;

  return frontMatter + originalNotice + content;
}

/**
 * Save post to file
 */
function savePost(article, contentData, outputDir) {
  const filename = sanitizeFilename(article.title || 'untitled') + '.md';
  const filepath = path.join(outputDir, filename);
  const content = createHexoPost(article, contentData);
  
  fs.writeFileSync(filepath, content, 'utf8');
  return filepath;
}

/**
 * Sleep helper
 */
function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Main function
 */
async function main() {
  console.log('='.repeat(60));
  console.log('CSDN to Hexo Blog Migration Tool');
  console.log('='.repeat(60));
  
  const fetchContent = process.argv.includes('--fetch-content');
  
  // Create output directory
  if (!fs.existsSync(CONFIG.outputDir)) {
    fs.mkdirSync(CONFIG.outputDir, { recursive: true });
  }
  
  // Step 1: Parse article list
  console.log('\n[Step 1] Parsing article list from "data" file...');
  const html = fs.readFileSync(CONFIG.inputFile, 'utf8');
  const articles = parseArticleList(html);
  console.log(`Found ${articles.length} articles`);
  
  // Save articles metadata to JSON
  fs.writeFileSync(CONFIG.articlesJson, JSON.stringify(articles, null, 2), 'utf8');
  console.log(`Saved article metadata to ${CONFIG.articlesJson}`);
  
  // Step 2: Process each article
  console.log('\n[Step 2] Processing articles...');
  if (fetchContent) {
    console.log('Note: Fetching full content from CSDN (this may take a while)...\n');
  } else {
    console.log('Note: Using metadata only. Run with --fetch-content to fetch full articles.\n');
  }
  
  let successCount = 0;
  let errorCount = 0;
  
  for (let i = 0; i < articles.length; i++) {
    const article = articles[i];
    const title = (article.title || 'Unknown').substring(0, 50);
    process.stdout.write(`[${i + 1}/${articles.length}] Processing: ${title}...`);
    
    try {
      let contentData = { content: '', tags: [], category: 'General' };
      
      if (fetchContent && article.url) {
        contentData = await fetchArticleContent(article.url);
        await sleep(CONFIG.delayBetweenRequests);
      }
      
      const filepath = savePost(article, contentData, CONFIG.outputDir);
      console.log(` ✓`);
      successCount++;
    } catch (e) {
      console.log(` ✗ Error: ${e.message}`);
      errorCount++;
      
      // Still save with metadata only on error
      try {
        savePost(article, { content: '', tags: [], category: 'General' }, CONFIG.outputDir);
      } catch (e2) {
        // ignore
      }
    }
  }
  
  // Summary
  console.log('\n' + '='.repeat(60));
  console.log('Migration Complete!');
  console.log('='.repeat(60));
  console.log(`Total articles: ${articles.length}`);
  console.log(`Successfully migrated: ${successCount}`);
  console.log(`Errors: ${errorCount}`);
  console.log(`\nPosts saved to: ${CONFIG.outputDir}/`);
  console.log(`Article metadata: ${CONFIG.articlesJson}`);
  console.log('\nNext steps:');
  console.log('1. Review the generated markdown files');
  console.log('2. Run "hexo generate" to build the site');
  console.log('3. Run "hexo server" to preview locally');
}

main().catch(console.error);

File: migrate_csdn.py
#!/usr/bin/env python3
"""
CSDN to Hexo Blog Migration Script
爬取CSDN文章完整内容并转换为Hexo博客格式
"""

import re
import os
import json
import time
import requests
from bs4 import BeautifulSoup
import html2text

# 配置
OUTPUT_DIR = "source/_posts"
ARTICLES_JSON = "articles.json"
DELAY = 2  # 请求间隔(秒)

HEADERS = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
    'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8',
    'Referer': 'https://blog.csdn.net/',
}

# HTML to Markdown 转换器
h2t = html2text.HTML2Text()
h2t.ignore_links = False
h2t.ignore_images = False
h2t.body_width = 0  # 不换行


def extract_urls_from_data(html_file):
    """从data文件提取所有文章URL和基本信息"""
    with open(html_file, 'r', encoding='utf-8') as f:
        html = f.read()
    
    soup = BeautifulSoup(html, 'html.parser')
    articles = []
    
    for block in soup.find_all('article', class_='blog-list-box'):
        article = {}
        
        # URL
        link = block.find('a', href=True)
        if link and '/article/details/' in link['href']:
            article['url'] = link['href']
            match = re.search(r'/article/details/(\d+)', article['url'])
            if match:
                article['id'] = match.group(1)
        
        # 标题
        h4 = block.find('h4')
        if h4:
            article['title'] = h4.get_text(strip=True)
        
        # 日期
        date_div = block.find('div', class_='view-time-box')
        if date_div:
            date_match = re.search(r'(\d{4})\.(\d{2})\.(\d{2})', date_div.get_text())
            if date_match:
                article['date'] = f"{date_match.group(1)}-{date_match.group(2)}-{date_match.group(3)}"
        
        # 阅读数
        views_div = block.find('div', class_='view-num-box')
        if views_div:
            m = re.search(r'(\d+)', views_div.get_text())
            if m:
                article['views'] = int(m.group(1))
        
        # 点赞数
        likes_div = block.find('div', class_='give-like-box')
        if likes_div:
            m = re.search(r'(\d+)', likes_div.get_text())
            if m:
                article['likes'] = int(m.group(1))
        
        # 评论和收藏
        comment_divs = block.find_all('div', class_='comment-box')
        if len(comment_divs) >= 1:
            m = re.search(r'(\d+)', comment_divs[0].get_text())
            if m:
                article['comments'] = int(m.group(1))
        if len(comment_divs) >= 2:
            m = re.search(r'(\d+)', comment_divs[1].get_text())
            if m:
                article['favorites'] = int(m.group(1))
        
        if article.get('url') and article.get('title'):
            articles.append(article)
    
    return articles


def fetch_article_detail(url):
    """爬取单篇文章的完整内容"""
    try:
        resp = requests.get(url, headers=HEADERS, timeout=30)
        resp.raise_for_status()
        resp.encoding = 'utf-8'
        soup = BeautifulSoup(resp.text, 'html.parser')
        
        result = {
            'content': '',
            'tags': [],
            'category': 'General',
            'cover': None
        }
        
        # 获取文章内容
        content_div = soup.find('div', id='content_views')
        if not content_div:
            content_div = soup.find('article', class_='baidu_pl')
        if not content_div:
            content_div = soup.find('div', class_='article_content')
        
        if content_div:
            # 移除不需要的元素
            for elem in content_div.find_all(['script', 'style', 'iframe']):
                elem.decompose()
            
            # 处理代码块，保留语言标识
            for pre in content_div.find_all('pre'):
                code = pre.find('code')
                if code:
                    lang = ''
                    if code.get('class'):
                        for cls in code['class']:
                            if 'language-' in cls or 'lang-' in cls:
                                lang = cls.replace('language-', '').replace('lang-', '')
                                break
                            if cls in ['python', 'java', 'javascript', 'js', 'cpp', 'c', 'html', 'css', 'sql', 'bash', 'shell', 'json', 'xml', 'yaml']:
                                lang = cls
                                break
                    code_text = code.get_text()
                    # 用特殊标记替换，后面转换
                    pre.replace_with(f'\n```{lang}\n{code_text}\n```\n')
            
            # 转换为Markdown
            html_content = str(content_div)
            result['content'] = h2t.handle(html_content)
            
            # 清理多余空行
            result['content'] = re.sub(r'\n{3,}', '\n\n', result['content'])
        
        # 获取标签 - 从blog-tags-box中查找tag-link-new
        tag_box = soup.find('div', class_='blog-tags-box')
        if tag_box:
            for tag_link in tag_box.find_all('a', class_='tag-link-new'):
                tag = tag_link.get_text(strip=True).lstrip('#')
                if tag and tag not in result['tags']:
                    result['tags'].append(tag)
        
        # 备用方法：从tags-box查找
        if not result['tags']:
            tag_box = soup.find('div', class_='tags-box')
            if tag_box:
                for tag_link in tag_box.find_all('a'):
                    tag = tag_link.get_text(strip=True).lstrip('#')
                    if tag and tag not in result['tags'] and len(tag) < 30:
                        result['tags'].append(tag)
        
        # 获取分类/专栏
        column_link = soup.find('a', class_='column_name')
        if column_link:
            cat = column_link.get_text(strip=True)
            if cat and cat not in ['原创', '转载']:
                result['category'] = cat
        
        # 获取封面图
        cover_img = soup.find('img', class_='article-cover')
        if not cover_img:
            # 尝试获取文章中第一张图片作为封面
            first_img = content_div.find('img') if content_div else None
            if first_img and first_img.get('src'):
                result['cover'] = first_img['src']
        else:
            result['cover'] = cover_img.get('src')
        
        return result
        
    except Exception as e:
        print(f"    错误: {e}")
        return {'content': '', 'tags': [], 'category': 'General', 'cover': None}


def sanitize_filename(title):
    """生成安全的文件名"""
    filename = re.sub(r'[<>:"/\\|?*]', '', title)
    filename = re.sub(r'\s+', '-', filename)
    filename = filename.strip('-')
    if len(filename) > 80:
        filename = filename[:80]
    return filename


def create_hexo_post(article, detail):
    """生成Hexo博客文章"""
    title = article.get('title', 'Untitled').replace('"', '\\"')
    date = article.get('date', '2023-01-01')
    category = detail.get('category', 'General')
    tags = detail.get('tags', [])
    content = detail.get('content', '')
    
    # Front matter
    front_matter = f'''---
title: "{title}"
date: {date}
updated: {date}
categories:
  - {category}
'''
    
    if tags:
        front_matter += 'tags:\n'
        for tag in tags:
            front_matter += f'  - {tag}\n'
    
    # CSDN统计数据
    front_matter += f'''csdn_views: {article.get('views', 0)}
csdn_likes: {article.get('likes', 0)}
csdn_comments: {article.get('comments', 0)}
csdn_favorites: {article.get('favorites', 0)}
csdn_url: {article.get('url', '')}
'''
    
    if detail.get('cover'):
        front_matter += f"cover: {detail['cover']}\n"
    
    front_matter += '---\n\n'
    
    # 原文信息提示
    notice = f'''> 本文迁移自CSDN博客
> 原文链接：[{article.get('title', '')}]({article.get('url', '')})
> 📊 {article.get('views', 0)} 阅读 | 👍 {article.get('likes', 0)} 点赞 | 💬 {article.get('comments', 0)} 评论 | ⭐ {article.get('favorites', 0)} 收藏

'''
    
    return front_matter + notice + content


def main():
    print("=" * 60)
    print("CSDN 博客迁移工具")
    print("=" * 60)
    
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    
    # 步骤1: 从data文件提取URL
    print("\n[步骤1] 从 data 文件提取文章URL...")
    articles = extract_urls_from_data('data')
    print(f"共发现 {len(articles)} 篇文章")
    
    # 保存元数据
    with open(ARTICLES_JSON, 'w', encoding='utf-8') as f:
        json.dump(articles, f, ensure_ascii=False, indent=2)
    
    # 步骤2: 爬取每篇文章
    print(f"\n[步骤2] 开始爬取文章内容 (间隔{DELAY}秒)...")
    
    success = 0
    failed = 0
    
    for i, article in enumerate(articles, 1):
        title = article.get('title', 'Unknown')[:40]
        print(f"[{i}/{len(articles)}] {title}...", end=' ')
        
        try:
            # 爬取文章详情
            detail = fetch_article_detail(article['url'])
            
            if detail['content']:
                # 生成并保存文章
                filename = sanitize_filename(article['title']) + '.md'
                filepath = os.path.join(OUTPUT_DIR, filename)
                
                post_content = create_hexo_post(article, detail)
                with open(filepath, 'w', encoding='utf-8') as f:
                    f.write(post_content)
                
                print(f"✓ ({len(detail['content'])}字)")
                success += 1
            else:
                print("✗ 内容为空")
                failed += 1
            
            # 延迟，避免被封
            time.sleep(DELAY)
            
        except Exception as e:
            print(f"✗ {e}")
            failed += 1
    
    # 总结
    print("\n" + "=" * 60)
    print("迁移完成!")
    print("=" * 60)
    print(f"成功: {success} 篇")
    print(f"失败: {failed} 篇")
    print(f"\n文章保存在: {OUTPUT_DIR}/")
    print("\n下一步:")
    print("  hexo generate")
    print("  hexo server")


if __name__ == '__main__':
    main()
